From dd0493612df82ec5e940573e8a10153c66eaf25b Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Tue, 22 Jul 2014 14:36:46 -0400
Subject: [virt] kvm: free resources after canceling async_pf

Message-id: <1406040016-3289-12-git-send-email-pbonzini@redhat.com>
Patchwork-id: 86090
O-Subject: [RHEL7 PATCH v2 011/221] kvm: free resources after canceling async_pf
Bugzilla: 1116936
RH-Acked-by: Radim Krcmar <rkrcmar@redhat.com>
RH-Acked-by: Bandan Das <bsd@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Radim Krcmar <rkrcmar@redhat.com>

When we cancel 'async_pf_execute()', we should behave as if the work was
never scheduled in 'kvm_setup_async_pf()'.
Fixes a bug when we can't unload module because the vm wasn't destroyed.

Signed-off-by: Radim Krcmar <rkrcmar@redhat.com>
Reviewed-by: Paolo Bonzini <pbonzini@redhat.com>
Reviewed-by: Gleb Natapov <gleb@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit 28b441e24088081c1e213139d1303b451a34a4f4)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/virt/kvm/async_pf.c b/virt/kvm/async_pf.c
index ea475cd..8a39dda 100644
--- a/virt/kvm/async_pf.c
+++ b/virt/kvm/async_pf.c
@@ -101,8 +101,11 @@ void kvm_clear_async_pf_completion_queue(struct kvm_vcpu *vcpu)
        typeof(*work), queue);
   cancel_work_sync(&work->work);
   list_del(&work->queue);
-  if (!work->done) /* work was canceled */
+  if (!work->done) { /* work was canceled */
+   mmdrop(work->mm);
+   kvm_put_kvm(vcpu->kvm); /* == work->vcpu->kvm */
    kmem_cache_free(async_pf_cache, work);
+  }
  }
 
  spin_lock(&vcpu->async_pf.lock);
-- 
1.7.1